### DISCUSSION TITLE
INFO: Howto connect to a device when discovery not possible - Solution

### DISCUSSION THREAD

**Discussion by roshanl** [ChristianTremblay/BAC0#487] [Show and tell]
1. **The challenge:** BAC0/Bacpypes3 running in a docker pod hosted on google cloud(IOT network) running on Redhat openshift. Pods are allocated IPS based on scaling, with a common host IP -  So whois/I am not possible. Have to communicate with MSTP behind a BACnet/IP router on local BMS network (OT network)
2. **The solution** - Gather all bacnet details using BACnet Explorer(Used Cimetrics explorer V4 to discover and save to XML/csv and imported to Excel) , use those data as device/object/MAC/network and create a device and the read properties from that device
3. **The working code is below.** Use as you wish if you have similar issues. THe below PC is not in docker but I use same code on the pod, just the device ID, local IP etc changes
4. **NOTE:** Change route_aware=true in device.json , otherwise bacnet through router to MSTP device will not work. Documentation on address format , about route_aware parameter can be found on bacpypes3 documentation.(if using in command line --route-aware as in python -m bacpypes3 --route-aware

import asyncio
import BAC0

async def main():
    # Define the BACnet device details
    device_id = 61105
    mac_address = '5'  # Ensure this is a string
    network_number = 95
    router_ip = '10.10.57.12'
    pc_ip = '10.10.57.220'
    remote_address=f'{network_number}:{mac_address}@{router_ip}' #https://github.com/JoelBender/BACpypes3/blob/48b8575b3fb91216acc474bb097e71caede577a4/doc/source/gettingstarted/addresses.rst
#route_aware address patterrn used for remote address

    # Initialize BAC0 with the PC's IP address
    bacnet = BAC0.connect(ip=pc_ip)

    # Manually add the device to the BACnet network
    device = await BAC0.device(device_id=device_id, address=f'{network_number}:{mac_address}@{router_ip}', network=f'{network_number}')

    # Now you can interact with the device
    # For example, reading a property
    property_value = await bacnet.read(f'{network_number}:{mac_address}@{router_ip} analog-value,104 present-value')
    print(f'Analog Value 104 Present Value: {property_value}')
#Can also do a readmultiple (Tested with an Alerton BMS and on an FCU controller)
_rpm = {'address': remote_address,
            'objects': {
                'analogValue:128': [' 'presentValue'],
                'analogValue:90': ['presentValue'],
                'analogValue:91': ['presentValue'],
                'analogValue:92': ['presentValue'],
                'analogValue:94': ['presentValue'],
                'analogValue:93': ['presentValue'],
                'analogValue:16': ['presentValue'],
                'analogValue:104': ['presentValue'],
                'analogValue:105': ['presentValue'],
                'binaryValue:5': ['presentValue'],
                'binaryValue:40': ['presentValue'],
                'analogOutput:0': ['presentValue'],
                'analogOutput:1': ['presentValue'],
                'binaryOutput:0': ['presentValue'],
                'binaryOutput:1': ['presentValue'],
                'binaryOutput:2': ['presentValue'],
                }
            }
            result = await bacnet.readMultiple('95:5@10.10.57.12', request_dict=_rpm)
            print(f'results: {result}')
# Run the main function within an event loop
if __name__ == "__main__":
    asyncio.run(main())

**Comment by ChristianTremblay** [comment]
I'm glad you found a solution. 
I would also like to let you know that I recently added a function to manually add a router to a network. This covers cases where the discovery process can't find a router but you know all the details.

## Use Router
```
address = "192.168.92.68"
network = "2001"
await bacnet.use_router((address, network))
```
By doing so, the app is now aware of this network and will make requests to the routers to get to it.

## Read vs Device
Regarding this part : 

```
# Manually add the device to the BACnet network
device = await BAC0.device(device_id=device_id, address=f'{network_number}:{mac_address}@{router_ip}', network=f'{network_number}')

# Now you can interact with the device
# For example, reading a property
property_value = await bacnet.read(f'{network_number}:{mac_address}@{router_ip} analog-value,104 present-value')
print(f'Analog Value 104 Present Value: {property_value}')
```
You don't need to create the `BAC0.device` if you use `await bacnet.read`

`BAC0.device` will allow you to create a "local version" of a device, with automatic polling, histories, quick access to objects, etc...

```
# Read value
device['pointName'] # gives last value read by polling
device['pointName'].value # forces a read on the network
device['pointName'] = 10 # write to the variable... if it's an output, make an override. If it's an input, make it out of service and write to presentValue, else, just write to it
device['pointName'].histories # guess what
# etc
```

Not creating the device will accelerate your app if you don't need it

**Comment by leanfm** [comment]
@ChristianTremblay As always, thank you for this great work!

I want to ask if the router is using a different port number, like 47814, and how to set up the router? When I tried the following, it always gave the error: ipaddress.AddressValueError: Only decimal digits permitted in '68:47814' in '192.168.92.68:47814'

address = "192.168.92.68:47814"
network = "2001"
await bacnet.use_router((address, network))

**Comment by ChristianTremblay** [comment]
I can't rememebr if 47814 is supported by bacpypes3.... I will have to check
